source/postprocess/geoid.cc

```diff
@@ -0,0 +1,754 @@
+                               this->get_fe(),
+                               quadrature_formula,
+                               update_values |
+                               update_gradients |
+                               update_q_points |
+                               update_JxW_values);
+
+      FEFaceValues<dim> fe_face_values (this->get_mapping(),
+                                        this->get_fe(),
+                                        quadrature_formula_face,
+                                        update_values |
+                                        update_gradients |
+                                        update_q_points |
+                                        update_JxW_values);
+
+      MaterialModel::MaterialModelInputs<dim> in(fe_values.n_quadrature_points, this->n_compositional_fields());
+      MaterialModel::MaterialModelOutputs<dim> out(fe_values.n_quadrature_points, this->n_compositional_fields());
+      MaterialModel::MaterialModelInputs<dim> in_face(fe_face_values.n_quadrature_points, this->n_compositional_fields());
+      MaterialModel::MaterialModelOutputs<dim> out_face(fe_face_values.n_quadrature_points, this->n_compositional_fields());
+
+      std::vector<std::vector<double> > composition_values (this->n_compositional_fields(),std::vector<double> (quadrature_formula.size()));
+      std::vector<std::vector<double> > face_composition_values (this->n_compositional_fields(),std::vector<double> (quadrature_formula_face.size()));
+
+      double integrated_surface_topography = 0;
+      double integrated_surface_area = 0;
+      double integrated_CMB_topography = 0;
+      double integrated_CMB_area = 0;
+      double integrated_top_layer_density = 0;
+      double integrated_top_layer_volume = 0;
+      double integrated_bottom_layer_density = 0;
+      double integrated_bottom_layer_volume = 0;
+
+      std::vector<std::pair<Point<dim>,std::pair<double,double> > > surface_stored_values;
+      std::vector<std::pair<Point<dim>,std::pair<double,double> > > CMB_stored_values;
+
+      // loop over all of the boundary cells and if one is at
+      // surface or CMB, evaluate the stress at its center
+      typename DoFHandler<dim>::active_cell_iterator
+      cell = this->get_dof_handler().begin_active(),
+      endc = this->get_dof_handler().end();
+
+      for (; cell!=endc; ++cell)
+        if (cell->is_locally_owned())
+          if (cell->at_boundary())
+            {
+              // see if the cell is at the *top* boundary or CMB, not just any boundary
+              unsigned int top_face_idx = numbers::invalid_unsigned_int;
+              unsigned int CMB_face_idx = numbers::invalid_unsigned_int;
+              {
+                for (unsigned int f=0; f<GeometryInfo<dim>::faces_per_cell; ++f)
+                  {
+                    if (cell->at_boundary(f) && this->get_geometry_model().depth (cell->face(f)->center()) < cell->face(f)->minimum_vertex_distance()/3)
```

Comment:
this logic seems rather complicated considering that you only support SphericalShell. Even without, a simple ``bool is_top_boundary`` is easier to use below.